R/teal_slice.R

```diff
@@ -268,22 +280,68 @@ c.teal_slice <- function(...) {
   ans
 }
 
+#' @param x `teal_slice` object
+#' @keywords internal
+to_list <- function(x){
+  checkmate::assert_class(x, "teal_slice")
+
+  x <- if (shiny::isRunning()) {
+    shiny::reactiveValuesToList(x)
+  } else {
+    shiny::isolate(shiny::reactiveValuesToList(x))
+  }
+
+  formals_names <- setdiff(names(formals(filter_var)), '...')
+
+  extra_names   <- setdiff(names(x), formals_names)
+  missing_names <- setdiff(formals_names, names(x))
+
+
+  x <- c(x, formals(filter_var)[missing_names])
+
+  x[c(formals_names, extra_names)]
+
+}
+
+#' @param x `list` object
+#' @param ... further arguments passed to [jsonlite::toJSON]
+#' @keywords internal
+to_json <- function(x, ...) {
+  checkmate::assert_class(x, "list")
+  jsonlite::toJSON(x, pretty = TRUE, auto_unbox = TRUE, digits = 16, ...)
+}
 
 # format method for teal_slice
 #' @export
+#' @param x `teal_slice` object
+#' @param show_all `logical(1)` should parameters set to NULL be returned
 #' @rdname teal_slice
 #' @keywords internal
 #'
-format.teal_slice <- function(x, show_all = FALSE, ...) {
+format.teal_slice <- function(x, show_all = FALSE, pretty = TRUE, ...) {
   checkmate::assert_flag(show_all)
+  checkmate::assert_flag(pretty)
 
-  x <- if (shiny::isRunning()) {
-    rev(shiny::reactiveValuesToList(x))
-  } else {
-    rev(shiny::isolate(shiny::reactiveValuesToList(x)))
-  }
+  x_json <- to_json(to_list(x))
+  x_json_c <- capture.output(print(x_json))
```

Comment:
So `format` calls `print` and then `print` calls `format`?
Also, I don't believe `format` should be capturing console output:
![image](https://github.com/insightsengineering/teal.slice/assets/114988527/e1eb7e37-0287-4d38-adff-f6aaa8f0c40b)

Reply:
@chlebowa but this is jsonlite:::print.JSON that is called here when calling print(x_json) - it's a trick to change JSON from a length 1 character to a character vector where each output line is an element of a vector so that further operations (removal of NULLs and centering with pretty_json) are possible

Reply:
`format` should only deal with strings and not involve input devices.

Reply:
alrighty @chlebowa but in `capture.output` there are no input devices in here (whathever that means). `capture.output` just catches the stuff that is passed to `cat`. We do the same in our tests. I am not writing to any file. I am dealing with a `character(1)` and `cat`` function that splits the output by '\n'

```{R}
> jsonlite::toJSON(data.frame(x = 5))
[{"x":5}] 
> jsonlite::toJSON(data.frame(x = 5), pretty = TRUE)
[
  {
    "x": 5
  }
] 
> jsonlite::toJSON(data.frame(x = 5), pretty = TRUE) -> z
> str(z)
 'json' chr "[\n  {\n    \"x\": 5\n  }\n]"
> print(z)
[
  {
    "x": 5
  }
] 
> jsonlite:::print.json
function (x, ...) 
{
    cat(x, "\n")
}
```

I can substitute `capture.output(print())` with strsplit(z, split = '\n')[[1]]

```{R}
[1] "["            "  {"          "    \"x\": 5" "  }"         
[5] "]"  
```

They are different in only one whitespace character at the end

![image](https://github.com/insightsengineering/teal.slice/assets/133694481/a5a99213-05f0-4381-a191-965f3463ad6a)

Reply:
0fb0664f5dfd268e020a11f4d8050c4fd6165c09

Reply:
My mistake, I mean output devices, sorry for the confusion.